server/request.go

```diff
@@ -392,44 +400,154 @@ func (r *RpcRequest) _writeRpcResponse(res *JsonRpcResponse) {
+		r.writeRpcResult(txHash)
 		return
 	}
 
-	param := make(map[string]string)
-	param["tx"] = r.rawTxHex
-	jsonRpcReq := NewJsonRpcRequest1(1, "eth_sendPrivateTransaction", param)
-	backendResp, respBytes, err := SendRpcWithSignatureAndParseResponse(r.relayUrl, r.relaySigningKey, jsonRpcReq)
+	sendPrivTxArgs := flashbotsrpc.FlashbotsSendPrivateTransactionRequest{Tx: r.rawTxHex}
+	_, err = FlashbotsRPC.FlashbotsSendPrivateTransaction(r.relaySigningKey, sendPrivTxArgs)
 	if err != nil {
-		r.logError("[sendTxToRelay] relay call failed for %s: %s - data: %s", txHash, err, *respBytes)
+		if errors.Is(err, flashbotsrpc.ErrRelayErrorResponse) {
+			r.log("[sendTxToRelay] relay error response: %v - rawTx: %s", err, r.rawTxHex)
+			r.writeRpcError(err.Error())
+		} else {
+			r.logError("[sendTxToRelay] relay call failed: %v - rawTx: %s", err, r.rawTxHex)
+			r.writeHeaderStatus(http.StatusInternalServerError)
+		}
+		return
+	}
+
+	r.writeRpcResult(txHash)
+	r.log("[sendTxToRelay] sent %s", txHash)
+}
+
+// Sends cancel-tx to relay as cancelPrivateTransaction, if initial tx was sent there too.
+func (r *RpcRequest) handleCancelTx() (requestCompleted bool) {
+	cancelTxHash := strings.ToLower(r.tx.Hash().Hex())
+	txFromLower := strings.ToLower(r.txFrom)
+	r.log("[cancel-tx] %s - check %s/%d", cancelTxHash, txFromLower, r.tx.Nonce())
+
+	// Get initial txHash by sender+nonce
+	initialTxHash, txHashFound, err := RState.GetTxHashForSenderAndNonce(txFromLower, r.tx.Nonce())
+	if err != nil {
+		r.logError("[cancel-tx] redis:GetTxHashForSenderAndNonce failed %v", err)
 		r.writeHeaderStatus(http.StatusInternalServerError)
+		return true
+	}
+
+	if !txHashFound { // not found, send to mempool
+		return false
+	}
+
+	// Check if initial tx was sent to relay
+	_, txWasSentToRelay, err := RState.GetTxSentToRelay(initialTxHash)
+	if err != nil {
+		r.logError("[cancel-tx] redis:GetTxSentToRelay failed: %s", err)
+		r.writeHeaderStatus(http.StatusInternalServerError)
+		return true
+	}
+
+	if !txWasSentToRelay { // was not sent to relay, send to mempool
+		return false
+	}
+
+	// Should send cancel-tx to relay. Check if cancel-tx was already sent before
+	_, cancelTxWasSentToRelay, err := RState.GetTxSentToRelay(cancelTxHash)
+	if err != nil {
+		r.logError("[cancel-tx] redis:GetTxSentToRelay error: %v", err)
+		r.writeHeaderStatus(http.StatusInternalServerError)
+		return true
+	}
+
+	if cancelTxWasSentToRelay { // already sent
+		r.writeRpcResult(cancelTxHash)
+		return true
+	}
+
+	r.log("[cancel-tx] sending to relay: %s for %s/%d", initialTxHash, txFromLower, r.tx.Nonce())
+
+	if DebugDontSendTx {
+		r.log("faked sending tx to relay, did nothing")
+		r.writeRpcResult(initialTxHash)
 		return
 	}
 
-	if backendResp.Error != nil {
-		r.logError("[sendTxToRelay] relay returned an error for %s: %s", txHash, backendResp.Error.Message)
+	cancelPrivTxArgs := flashbotsrpc.FlashbotsCancelPrivateTransactionRequest{TxHash: initialTxHash}
```

Comment:
Just a reminder to merge the private tx updates on your flashbots repo

Reply:
👍 

https://github.com/metachris/flashbotsrpc/pull/4